metadata-ingestion/tests/unit/test_transform_dataset.py

```diff
@@ -2665,3 +2671,167 @@ def test_pattern_dataset_schema_tags_transformation_patch(
+
+    first_data_product_aspect = json.loads(
+        outputs[3].record.aspect.value.decode("utf-8")
+    )
+    assert len(first_data_product_aspect) == 2
+    assert first_data_product_aspect[0]["value"][
+        "destinationUrn"
+    ] == builder.make_dataset_urn("bigquery", "example1")
+    assert first_data_product_aspect[1]["value"][
+        "destinationUrn"
+    ] == builder.make_dataset_urn("bigquery", "example3")
+
+    second_data_product_aspect = json.loads(
+        outputs[4].record.aspect.value.decode("utf-8")
+    )
+    assert len(second_data_product_aspect) == 1
+    assert second_data_product_aspect[0]["value"][
+        "destinationUrn"
+    ] == builder.make_dataset_urn("bigquery", "example2")
+
+    assert isinstance(outputs[5].record, EndOfStream)
+
+
+def test_pattern_dataset_data_product_transformation(mock_time):
+    transformer = PatternAddDatasetDataProduct.create(
+        {
+            "dataset_to_data_product_urns_pattern": {
+                "rules": {
+                    ".*example1.*": "urn:li:dataProduct:first",
+                    ".*": "urn:li:dataProduct:second",
+                }
+            },
+        },
+        PipelineContext(run_id="test-dataproducts"),
+    )
+
+    outputs = list(
+        transformer.transform(
+            [
+                RecordEnvelope(input, metadata={})
+                for input in [
+                    make_generic_dataset(
+                        entity_urn=builder.make_dataset_urn("bigquery", "example1")
+                    ),
+                    make_generic_dataset(
+                        entity_urn=builder.make_dataset_urn("bigquery", "example2")
+                    ),
+                    make_generic_dataset(
+                        entity_urn=builder.make_dataset_urn("bigquery", "example3")
+                    ),
+                    EndOfStream(),
+                ]
+            ]
+        )
+    )
+
+    assert len(outputs) == 6
+
+    # Check new dataproduct entity should be there
+    assert outputs[3].record.entityType == "dataProduct"
+    assert outputs[3].record.entityUrn == "urn:li:dataProduct:first"
+    assert outputs[3].record.aspectName == "dataProductProperties"
+
+    first_data_product_aspect = json.loads(
+        outputs[3].record.aspect.value.decode("utf-8")
+    )
+    assert len(first_data_product_aspect) == 1
+    assert first_data_product_aspect[0]["value"][
+        "destinationUrn"
+    ] == builder.make_dataset_urn("bigquery", "example1")
+
+    second_data_product_aspect = json.loads(
+        outputs[4].record.aspect.value.decode("utf-8")
+    )
+    assert len(second_data_product_aspect) == 2
+    assert second_data_product_aspect[0]["value"][
+        "destinationUrn"
+    ] == builder.make_dataset_urn("bigquery", "example2")
+    assert second_data_product_aspect[1]["value"][
+        "destinationUrn"
+    ] == builder.make_dataset_urn("bigquery", "example3")
```

Comment:
this is much simpler

```suggestion
    assert [item["value"]["destinationUrn"] for item in second_data_product_aspect] == [builder.make_dataset_urn("bigquery", "example2"), builder.make_dataset_urn("bigquery", "example3")]
```